validator/sawtooth_validator/execution/scheduler_parallel.py

```diff
@@ -116,23 +116,25 @@ def find_write_predecessors(self, address):
         # Children readers must be added, since their reads must happen prior
         # to the write.
 
-        predecessors = set()
-
         tokens = self._tokenize_address(address)
 
-        # First, find enclosing_writer, enclosing_readers, and address_node (if
-        # the tree is deep enough).
-        enclosing_writer = None
-        address_node = None
+        predecessors = set()
+
+        # First, walk down from the root to the address, collecting all readers
+        # and updating the enclosing_writer if needed.
 
         node = self._root
-        for i, token in enumerate(tokens):
+        enclosing_writer = node.writer  # possibly None
+
+        predecessors.update(set(node.readers))
+
+        for token in tokens:
             if token not in node.children:
-                break
+                if enclosing_writer is not None:
+                    predecessors.add(enclosing_writer)
+                return predecessors
 
             node = node.children[token]
-            if i == len(tokens) - 1:
```

Comment:
If I understand this correctly, it looks like node will equal the address node when we leave the for loop, which is guaranteed because of the return within the loop would have prevented partial addresses from continuing to be processed.  Might be worth a comment that this side-effect of the return is very important to ensuring correctness.  If we were not returning, then we would process a node higher in the tree which would be very incorrect.